plugins/inputs/exec/exec.go

```diff
@@ -105,43 +84,46 @@ func (c CommandRunner) Run(
 	cmd.Stdout = &out
 	cmd.Stderr = &stderr
 
-	if err := internal.RunTimeout(cmd, e.Timeout.Duration); err != nil {
-		switch e.parser.(type) {
-		case *nagios.NagiosParser:
-			AddNagiosState(err, acc)
-		default:
-			var errMessage = ""
-			if stderr.Len() > 0 {
-				stderr = removeCarriageReturns(stderr)
-				// Limit the number of bytes.
-				didTruncate := false
-				if stderr.Len() > MaxStderrBytes {
-					stderr.Truncate(MaxStderrBytes)
+	_, isNagiosParser := e.parser.(*nagios.NagiosParser)
+
+	err = internal.RunTimeout(cmd, e.Timeout.Duration)
+	nagiosErr := err
+	if err != nil && !isNagiosParser {
+		var errMessage = ""
+		if stderr.Len() > 0 {
+			stderr = removeCarriageReturns(stderr)
+			// Limit the number of bytes.
+			didTruncate := false
+			if stderr.Len() > MaxStderrBytes {
+				stderr.Truncate(MaxStderrBytes)
+				didTruncate = true
+			}
+			if i := bytes.IndexByte(stderr.Bytes(), '\n'); i > 0 {
+				// Only show truncation if the newline wasn't the last character.
+				if i < stderr.Len()-1 {
 					didTruncate = true
 				}
-				if i := bytes.IndexByte(stderr.Bytes(), '\n'); i > 0 {
-					// Only show truncation if the newline wasn't the last character.
-					if i < stderr.Len()-1 {
-						didTruncate = true
-					}
-					stderr.Truncate(i)
-				}
-				if didTruncate {
-					stderr.WriteString("...")
-				}
-
-				errMessage = fmt.Sprintf(": %s", stderr.String())
+				stderr.Truncate(i)
 			}
-			return nil, fmt.Errorf("exec: %s for command '%s'%s", err, command, errMessage)
-		}
-	} else {
-		switch e.parser.(type) {
-		case *nagios.NagiosParser:
-			AddNagiosState(nil, acc)
+			if didTruncate {
+				stderr.WriteString("...")
+			}
+
+			errMessage = fmt.Sprintf(": %s", stderr.String())
 		}
+		return nil, fmt.Errorf("exec: %s for command '%s'%s", err, command, errMessage)
 	}
 
 	out = removeCarriageReturns(out)
+
+	if isNagiosParser {
+		exitCode, err := nagios.GetExitCode(nagiosErr)
+		if err != nil {
+			return nil, fmt.Errorf("exec: get exit code: %s", err)
+		}
+		nagios.AppendExitCode(&out, exitCode)
```

Comment:
I feel like the `state`, `service_output`, `long_service_output` semantically belong together and should be a part of the same measurement.

The current parser interface doesn't allow to pass any extra information to the parser, whilst the `service_output` and the `long_service_output` are known during the parsing time only.

The parser will properly parse bytes for which the `nagios.AppendExitCode(...)` wasn't called, it will assume the default state to be 0.

Reply:
It's not my primary concern but wouldn't the state be set incorrectly depending on how the output ends?  Perhaps we could add the state to the `telegraf.Metric` after it has been returned from the parser?

Reply:
- Because we encode the state in the `exec` plugin, no matter how the output ends, the parser will always properly decode the state and strip those additional bytes.

  There might be an issue, when the parser is used somewhere else. And it is my bad, because I did not properly document the behavior of the `Parse` method. I will add the docs, but first, could you please comment on the next point:
- Appending the error code after we return from the parser is definitely possible. But will require me to do something like this:

```golang

// To return the exit code through the error handling mechanism.
//
type exitCode int

func (c exitCode) Error() string {}

// The callee changes
//
func (c CommandRunner) Run(
	e *Exec,
	command string,
	acc telegraf.Accumulator,
) ([]byte, error) {
	<...irrelevant...>

	out = <output from the check>

	if isNagiosParser {
		exitCode, err := nagios.GetExitCode(nagiosErr)
		if err != nil {
			return nil, fmt.Errorf("exec: get exit code: %s", err)
		}
		return out.Bytes(), exitCode(exitCode)
	}

	return out.Bytes(), nil
}

// The caller changes
//
func (e *Exec) ProcessCommand(command string, acc telegraf.Accumulator, wg *sync.WaitGroup) {
	defer wg.Done()

	out, err := e.runner.Run(e, command, acc)
        var exitCode int
	if err != nil {
		if v, ok := err.(exitCode); ok {
			exitCode = int(v)
		} else {
			acc.AddError(err)
			return
		}
	}
      

	metrics, err := e.parser.Parse(out)
	if err != nil {
		acc.AddError(err)
		return
	}
	
        if _, ok := e.parser.(*nagios.NagiosParser); ok {
                // Find the metric with name == "nagios_state" and append the "state" field to it.
                // Can be a convention that it is the last metric, or iterate through all and find it.
               m := <somehow find the metric>
               m.AddField("state", exitCode)
        }
        
        for _, metric := range metrics {
		acc.AddFields(metric.Name(), metric.Fields(), metric.Tags(), metric.Time())
	}
}
```
- I am relying on your expertise here. Which approach seem to be better for you?

Reply:
@danielnelson, what is your take on this ^^?

Reply:
To be sure, neither solution is great, and if I could do it over then I would probably make nagios it's own plugin separate from exec and scrap the standalone parser.

Putting that aside though, since I don't think it is worth a breaking change... I would parse, then if this is the NagiosParser, get the error code and iterate over the metrics looking for the `nagios_state` metric using `metric.Name()`.  You will also want to watch for the case where the state metric is not found, and add the field as a new metric.

One more thing, could you do me a favor and use this form for adding the metrics?
```
        for _, metric := range metrics {
		acc.AddMetric(metric)
	}
```